bpf-gpl/conntrack.h

```diff
@@ -200,7 +208,40 @@ static CALI_BPF_INLINE int calico_ct_v4_create_tracking(struct ct_create_ctx *ct
 		CALI_DEBUG("CT-ALL Whitelisted dest side - to EP\n");
 	}
 
-	err = cali_v4_ct_update_elem(k, &ct_value, 0);
+	err = cali_v4_ct_update_elem(k, &ct_value, BPF_NOEXIST);
+
+	if (err == -17 /* EEXIST */) {
+		CALI_DEBUG("Source collision for 0x%x:%d\n", bpf_htonl(ip_src), sport);
+
+		ct_ctx->orig_sport = ct_value.orig_sport = sport;
+
+		sport = psnat_get_port();
+		CALI_DEBUG("New sport %d\n", sport);
+
+		bool srcLTDest = (ip_src < ip_dst) || ((ip_src == ip_dst) && sport < dport);
+
+		if (srcLTDest) {
+			*k = (struct calico_ct_key) {
+				.protocol = ct_ctx->proto,
+					.addr_a = ip_src, .port_a = sport,
+					.addr_b = ip_dst, .port_b = dport,
+			};
+		} else  {
+			*k = (struct calico_ct_key) {
+				.protocol = ct_ctx->proto,
+					.addr_a = ip_dst, .port_a = dport,
+					.addr_b = ip_src, .port_b = sport,
+			};
+		}
+
+		dump_ct_key(k);
+
+		if ((err = cali_v4_ct_update_elem(k, &ct_value, BPF_NOEXIST))) {
```

Comment:
I wonder if we should try calling psnat_get_port again?  Maybe we were just unlucky.

Reply:
yes, we can be unlucky, then we drop the packet. However, it will try to send another SYN (up to 3?) so it may pick a new port then. The connection may time out, but hat is still better than breaking the existing connection.

I am not sure if we can easily keep calling it - that is what I would do outside of BPF :sweat_smile: 

We were considering sending back and RST to avoid the timeout. I just did not get to the RST part yet. TBH I would like to see how the random pick behaves in the wild firts - but there is no harm in adding the RST, except that middle-boxes shouldn't send one :sweat_smile:

Reply:
I added a test that demonstrates that another collision is possible and that a retransmission is likely to fix this.